codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -0,0 +1,438 @@
+     * @param isErrorCodeInBody A boolean that indicates if the error code for the implementing protocol is located in
+     *                          the error response body, meaning this generator will parse the body before attempting to
+     *                          load an error code.
+     */
+    public HttpBindingProtocolGenerator(boolean isErrorCodeInBody) {
+        this.isErrorCodeInBody = isErrorCodeInBody;
+    }
+
+    @Override
+    public ApplicationProtocol getApplicationProtocol() {
+        return ApplicationProtocol.createDefaultHttpApplicationProtocol();
+    }
+
+    @Override
+    public void generateSharedComponents(GenerationContext context) {
+        // pass
+    }
+
+    @Override
+    public void generateRequestSerializers(GenerationContext context) {
+        TopDownIndex topDownIndex = context.getModel().getKnowledge(TopDownIndex.class);
+
+        Set<OperationShape> containedOperations = new TreeSet<>(
+                topDownIndex.getContainedOperations(context.getService()));
+        for (OperationShape operation : containedOperations) {
+            OptionalUtils.ifPresentOrElse(
+                    operation.getTrait(HttpTrait.class),
+                    httpTrait -> generateOperationSerializer(context, operation),
+                    () -> LOGGER.warning(String.format(
+                            "Unable to generate %s protocol request bindings for %s because it does not have an "
+                                    + "http binding trait", getProtocol(), operation.getId())));
+        }
+    }
+
+    @Override
+    public void generateResponseDeserializers(GenerationContext context) {
+        // pass
+    }
+
+    /**
+     * Gets the default serde format for timestamps.
+     *
+     * @return Returns the default format.
+     */
+    protected abstract Format getDocumentTimestampFormat();
+
+    /**
+     * Gets the default content-type when a document is synthesized in the body.
+     *
+     * @return Returns the default content-type.
+     */
+    protected abstract String getDocumentContentType();
+
+    private void generateOperationSerializer(
+            GenerationContext context,
+            OperationShape operation
+    ) {
+        generateOperationHttpBindingSerializer(context, operation);
+    }
+
+    private boolean isRestBinding(HttpBinding.Location location) {
+        return location == HttpBinding.Location.HEADER
+                || location == HttpBinding.Location.PREFIX_HEADERS
+                || location == HttpBinding.Location.LABEL
+                || location == HttpBinding.Location.QUERY;
+    }
+
+    private void generateOperationHttpBindingSerializer(
+            GenerationContext context,
+            OperationShape operation
+    ) {
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Model model = context.getModel();
+        GoWriter writer = context.getWriter();
+
+        Shape inputShape = model.getShape(operation.getInput()
```

Comment:
You could either do `expectShape` here or `flatMap` the two optionals together so you only have to throw an exception once.